### PR TITLE
fix: inject new set of tokens when the session is resumed

### PR DESCRIPTION
**Pull request by olevski**
This fixes the issues we have had with expiring Gitlab access tokens.

/deploy renku=fix-gitlab-token-in-sessions

### CODE REVIEW

File: renku_notebooks/api/classes/k8s_client.py [hunk 1]
```diff
@@ -183,6 +188,140 @@ def list_servers(self, label_selector: Optional[str] = None) -> List[Dict[str, A
...
+        git_proxy_container = ss.spec.template.spec.containers[git_proxy_container_index]
+        git_init_container_index = 2
+        git_init_container = ss.spec.template.spec.init_containers[git_init_container_index]
+        patch = []
+        expires_at_env = find_env_var(git_proxy_container, "GITLAB_OAUTH_TOKEN_EXPIRES_AT")
+        gitlab_token_env = find_env_var(git_proxy_container, "GITLAB_OAUTH_TOKEN")
+        git_init_token_env = find_env_var(git_init_container, "GIT_CLONE_USER__OAUTH_TOKEN")
+        renku_access_token_env = find_env_var(git_proxy_container, "RENKU_ACCESS_TOKEN")
+        renku_refresh_token_env = find_env_var(git_proxy_container, "RENKU_REFRESH_TOKEN")
+        if not any(
```
Comment (Panaetius) [inline renku_notebooks/api/classes/k8s_client.py:263]: should this be `all`? now it'd pass if a single one is there

File: renku_notebooks/api/classes/k8s_client.py [hunk 2]
```diff
@@ -353,6 +492,15 @@ def delete_server(self, server_name: str, safe_username: str, forced: bool = Fal
...
             self.session_ns_client.delete_server(server_name, forced)
 
+    def patch_tokens(self, server_name, renku_tokens: RenkuTokens, gitlab_token: GitlabToken):
+        """Patch the Renku and Gitlab access tokens used in a session."""
+        if self.session_ns_client:
+            self.session_ns_client.patch_statefulset_tokens(server_name, renku_tokens, gitlab_token)
+            self.session_ns_client.patch_image_pull_secret(server_name, gitlab_token)
+        else:
+            self.renku_ns_client.patch_statefulset_tokens(server_name, renku_tokens, gitlab_token)
+            self.renku_ns_client.patch_image_pull_secret(server_name, gitlab_token)
```
Comment (Panaetius) [inline renku_notebooks/api/classes/k8s_client.py:502]:
```suggestion
        if self.session_ns_client:
            client = self.session_ns_client
        else:
            client = self.renku_ns_client
        client.patch_statefulset_tokens(server_name, renku_tokens, gitlab_token)
        client.patch_image_pull_secret(server_name, gitlab_token)
```

**Review by Panaetius — CHANGES_REQUESTED** [review]

**Review by mohammad-alisafaee — DISMISSED** [review]
LGTM, thanks!

File: renku_notebooks/api/classes/k8s_client.py (hunk 1 above)
Comment (olevski) [reply renku_notebooks/api/classes/k8s_client.py:263]: Yes you are correct.

**Review by Panaetius — APPROVED** [review]

**Pull request merged**